File: ios/NitroGoogleCast/Converters/MediaQueueItem+toGckMediaQueueItem.swift
import Foundation
import GoogleCast
import NitroModules

/// Converts a generated `MediaQueueItem` struct into a `GCKMediaQueueItem` via
/// `GCKMediaQueueItemBuilder`.
///
/// Reverse lives in `GCKMediaQueueItem+toMediaQueueItem.swift`. Notes:
/// - `itemId` is assigned by the receiver and has NO builder setter, so it is dropped on the
///   forward trip and always comes back `nil` (cross-platform: Android must reconcile).
/// - `autoplay` is non-optional in GCK; when absent, GCK's builder default applies and the
///   reverse always emits a value (the corpus pins the observed default).
/// - Time intervals left unset use GCK's sentinel, which the reverse maps back to `nil`.
extension MediaQueueItem {
  func toGckMediaQueueItem() -> GCKMediaQueueItem {
    let builder = GCKMediaQueueItemBuilder()
    builder.mediaInformation = mediaInfo.toGckMediaInformation()
    if let autoplay { builder.autoplay = autoplay }
    if let startTime { builder.startTime = startTime }
    if let playbackDuration { builder.playbackDuration = playbackDuration }
    if let preloadTime { builder.preloadTime = preloadTime }
    if let activeTrackIds {
      builder.activeTrackIDs = activeTrackIds.map { NSNumber(value: $0) }
    }
    if let customData { builder.customData = customData.toGckCustomData() }
    return builder.build()
  }
}

File: android/src/main/java/com/margelo/nitro/googlecast/converters/MediaQueueItem+toGckMediaQueueItem.kt
package com.margelo.nitro.googlecast.converters

import com.margelo.nitro.googlecast.MediaQueueItem
import com.google.android.gms.cast.MediaQueueItem as GckMediaQueueItem

/**
 * Converts a generated [MediaQueueItem] struct into a Google Cast [GckMediaQueueItem].
 *
 * The reverse lives in `GckMediaQueueItem+toMediaQueueItem.kt`. Unlike most GCK time fields,
 * MediaQueueItem's `startTime` / `playbackDuration` / `preloadTime` are already SECONDS
 * (`double`) in GCK Android, so they pass through without scaling.
 */
internal fun MediaQueueItem.toGckMediaQueueItem(): GckMediaQueueItem {
  val builder = GckMediaQueueItem.Builder(mediaInfo.toGckMediaInfo())
  itemId?.let { builder.setItemId(it.toInt()) }
  activeTrackIds?.let { ids -> builder.setActiveTrackIds(LongArray(ids.size) { ids[it].toLong() }) }
  autoplay?.let { builder.setAutoplay(it) }
  playbackDuration?.let { builder.setPlaybackDuration(it) }
  preloadTime?.let { builder.setPreloadTime(it) }
  startTime?.let { builder.setStartTime(it) }
  customData?.let { builder.setCustomData(it.toJsonObject()) }
  return builder.build()
}
